File: PhotoPicker/src/main/java/hai/ithust/photopicker/utils/MediaStoreHelper.java
package hai.ithust.photopicker.utils;

import android.content.Context;
import android.database.Cursor;
import android.provider.MediaStore;

import java.util.ArrayList;
import java.util.List;

import hai.ithust.photopicker.entity.PhotoDirectory;

import static android.provider.BaseColumns._ID;
import static android.provider.MediaStore.Images.ImageColumns.BUCKET_DISPLAY_NAME;
import static android.provider.MediaStore.Images.ImageColumns.BUCKET_ID;
import static android.provider.MediaStore.MediaColumns.DATA;
import static android.provider.MediaStore.MediaColumns.DATE_ADDED;
import static android.provider.MediaStore.MediaColumns.MIME_TYPE;
import static android.provider.MediaStore.MediaColumns.SIZE;

/**
 * @author conghai on 12/20/18.
 */
public class MediaStoreHelper {
    public static List<PhotoDirectory> getPhotoDirs(Context context) {
        List<PhotoDirectory> directories = new ArrayList<>();

        String[] protections = {
                MediaStore.Images.Media._ID,
                MediaStore.Images.Media.DATA,
                MediaStore.Images.Media.BUCKET_ID,
                MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
                MediaStore.Images.Media.DATE_ADDED,
                MediaStore.Images.Media.SIZE
        };
        Cursor data = context.getApplicationContext().getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                , protections
                , MIME_TYPE + "=? or " + MIME_TYPE + "=? or " + MIME_TYPE + "=? " + ""
                , new String[]{"image/jpeg", "image/png", "image/jpg"}
                , MediaStore.Images.Media.DATE_ADDED + " DESC");
        if (data != null) {
            while (data.moveToNext()) {
                int imageId = data.getInt(data.getColumnIndexOrThrow(_ID));
                String bucketId = data.getString(data.getColumnIndexOrThrow(BUCKET_ID));
                String name = data.getString(data.getColumnIndexOrThrow(BUCKET_DISPLAY_NAME));
                String path = data.getString(data.getColumnIndexOrThrow(DATA));
                long size = data.getInt(data.getColumnIndexOrThrow(SIZE));

                if (size < 1) continue;

                PhotoDirectory photoDirectory = new PhotoDirectory();
                photoDirectory.setId(bucketId);
                photoDirectory.setName(name);

                if (!directories.contains(photoDirectory)) {
                    photoDirectory.setCoverPath(path);
                    photoDirectory.addPhoto(imageId, path);
                    photoDirectory.setDateAdded(data.getLong(data.getColumnIndexOrThrow(DATE_ADDED)));
                    directories.add(photoDirectory);
                } else {
                    directories.get(directories.indexOf(photoDirectory)).addPhoto(imageId, path);
                }
            }
            data.close();
        }
        return directories;
    }
}

File: PhotoPicker/src/main/java/hai/ithust/photopicker/utils/MediaStoreHelper.kt
package hai.ithust.photopicker.utils

import android.content.ContentUris
import android.content.Context
import android.provider.BaseColumns
import android.provider.MediaStore
import android.provider.MediaStore.MediaColumns
import hai.ithust.photopicker.entity.PhotoDirectory
import java.util.ArrayList

/**
 * @author conghai on 12/20/18.
 */
object MediaStoreHelper {
    @JvmStatic
    fun getPhotoDirs(context: Context): List<PhotoDirectory> {
        val directories: MutableList<PhotoDirectory> = mutableListOf()
        val protections = arrayOf(
                MediaStore.Images.Media._ID,
                MediaStore.Images.Media.BUCKET_ID,
                MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
                MediaStore.Images.Media.SIZE
        )
        val data = context.contentResolver.query(
                MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                protections,
                MediaColumns.MIME_TYPE + "=? or " + MediaColumns.MIME_TYPE + "=? or " + MediaColumns.MIME_TYPE + "=? " + "",
                arrayOf("image/jpeg", "image/png", "image/jpg"),
                MediaStore.Images.Media.DATE_ADDED + " DESC"
        )
        if (data != null) {
            while (data.moveToNext()) {
                val size = data.getInt(data.getColumnIndexOrThrow(MediaColumns.SIZE)).toLong()
                if (size < 1) continue

                val imageId = data.getInt(data.getColumnIndexOrThrow(BaseColumns._ID))
                val bucketId = data.getString(data.getColumnIndexOrThrow(MediaColumns.BUCKET_ID))
                val name = data.getString(data.getColumnIndexOrThrow(MediaColumns.BUCKET_DISPLAY_NAME))
                val photoUri = ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, imageId.toLong())
                val index = directories.indexOfFirst { it.id == bucketId }
                if (index in directories.indices) {
                    directories[index].addPhoto(photoUri)
                } else {
                    directories.add(PhotoDirectory(bucketId, photoUri, name))
                }
            }
            data.close()
        }
        return directories
    }
}